pkg/cloudproduct/gke/gke_test.go

```diff
@@ -66,17 +66,201 @@ func TestSyncPodPortsToGameServer(t *testing.T) {
 				return
 			}
 			if assert.NoError(t, err) {
-				if diff := cmp.Diff(tc.wantGS, tc.gs); diff != "" {
-					t.Errorf("GameServer diff (-want +got):\n%s", diff)
-				}
-				if diff := cmp.Diff(oldPod, tc.pod); diff != "" {
-					t.Errorf("Pod was modified (-old +new):\n%s", diff)
-				}
+				require.Equal(t, tc.wantGS, tc.gs)
+				require.Equal(t, oldPod, tc.pod)
 			}
 		})
 	}
 }
 
+func TestValidateGameServer(t *testing.T) {
+	for name, tc := range map[string]struct {
+		gs   *agonesv1.GameServer
+		want []metav1.StatusCause
+	}{
+		"no ports => validated": {
+			gs: &agonesv1.GameServer{},
+		},
+		"good ports => validated": {
+			gs: &agonesv1.GameServer{
+				Spec: agonesv1.GameServerSpec{
+					Ports: []agonesv1.GameServerPort{
+						{
+							Name:          "some-tcpudp",
+							PortPolicy:    agonesv1.Dynamic,
+							ContainerPort: 4321,
+							Protocol:      agonesv1.ProtocolTCPUDP,
+						},
+						{
+							Name:          "awesome-udp",
+							PortPolicy:    agonesv1.Dynamic,
+							ContainerPort: 1234,
+							Protocol:      corev1.ProtocolUDP,
+						},
+						{
+							Name:          "awesome-tcp",
+							PortPolicy:    agonesv1.Dynamic,
+							ContainerPort: 1234,
+							Protocol:      corev1.ProtocolTCP,
+						},
+					},
+				},
+			},
+		},
+		"bad policy => fails validation": {
+			gs: &agonesv1.GameServer{
+				Spec: agonesv1.GameServerSpec{
+					Ports: []agonesv1.GameServerPort{
+						{
+							Name:          "best-tcpudp",
+							PortPolicy:    agonesv1.Dynamic,
+							ContainerPort: 4321,
+							Protocol:      agonesv1.ProtocolTCPUDP,
+						},
+						{
+							Name:          "bad-udp",
+							PortPolicy:    agonesv1.Static,
+							ContainerPort: 1234,
+							Protocol:      corev1.ProtocolUDP,
+						},
+						{
+							Name:          "another-bad-udp",
+							PortPolicy:    agonesv1.Static,
+							ContainerPort: 1234,
+							Protocol:      corev1.ProtocolUDP,
+						},
+					},
+				},
+			},
+			want: []metav1.StatusCause{
+				{
+					Type:    "FieldValueInvalid",
+					Message: "PortPolicy must be Dynamic on GKE Autopilot",
+					Field:   "bad-udp.portPolicy",
+				},
+				{
+					Type:    "FieldValueInvalid",
+					Message: "PortPolicy must be Dynamic on GKE Autopilot",
+					Field:   "another-bad-udp.portPolicy",
+				},
+			},
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			causes := (&gkeAutopilot{}).ValidateGameServer(tc.gs)
+			require.Equal(t, tc.want, causes)
+		})
+	}
+}
+
+func TestAutopilotPortAllocator(t *testing.T) {
+	for name, tc := range map[string]struct {
+		gs     *agonesv1.GameServer
```

Comment:
As on the previous one, it might be simpler to pass in a slice ports and specify the slice of ports you expect to get back. In this case, there would also need to be a bool to verify whether the annotation is set on the resulting game server. 

Same question as on the last one - is 2 test cases enough to catch all edge cases?

Reply:
I looked longer at this one as it's more complex, and added one other minor test that adds a TCPUDP after the normal ports. One thing is that the "2 test cases" are really compound .. I layered several pieces of complexity into the success case. Again, checking coverage, the coverage here is solid, with the exception of an error from `json.Marshal`. I have historically never bothered testing such errors, as they can be more work to induce than is worth it.